fetch.bs

```diff
@@ -3270,6 +3281,109 @@ through TLS using ALPN. The protocol cannot be spoofed through HTTP requests in
 </div>
 
 
+<h3 id=https-upgrading>HTTPS upgrading</h3>
+
+<p>User agents may optionally upgrade requests with URLs that are not
+<a>potentially trustworthy URLs</a> to attempt to fetch them over
+<a>potentially trustworthy URLs</a>. If an upgraded request fails with a network error, it is
+retried over the original URL.
+
+<p>The HTTPS upgrading algorithm consists of <a>upgrade an HTTP request</a> and
+<a>HTTPS upgrade fallback</a> algorithms.
+
+<h4 id=https-upgrading-upgrade>HTTPS upgrade algorithm</h4>
+
+<div algorithm>
+<p>To <dfn>upgrade an HTTP request</dfn> given a <a for=/>request</a> <var>request</var>:
+
+<ol>
+ <li>
+  <p>If one or more of the following conditions are met, return:
+  <ul>
+   <li><p><var>request</var>'s <a for="request">destination</a> is not "<code>document</code>"
+
+   <li><p><var>request</var>'s <a for="request">method</a> is not "<code>GET</code>"
+
+   <li><p><var>request</var>'s <a for="request">URL</a>'s <a for="url">scheme</a> is not
+   "<code>http</code>"
+
+   <li><p><var>request</var>'s <a for="request">URL</a>'s <a for="url">host</a> is exempted from
+   upgrades in an <a>implementation-defined</a> way.
```

Comment:
Maybe we can give some examples here. Also, is this purely based on host or is port part of the consideration? Maybe we should talk about origin instead?

Reply:
I added an example in the examples section below for an allowlisting case. Happy to add more.

Speaking of allowlisting: It's currently hostname-based in Chrome for a few reasons:
- It matched how certificate error decisions were stored for SSL interstitials
- Making it hostname based, thus overly inclusive, reduced the risk of breakage. Failing an upgrade once opts out the entire site from upgrades again for the allowlisting duration. 
- The allowlist is shared with Chrome's HTTPS-First Mode which shows an interstitial before falling back to HTTP. So making it more fine grained could result in more interstitials and lead to warning fatigue.

That said, I think we can relax this and give more leeway to UAs. If we say origin here, but the UA still decides to exempt based on hostname only, would that be acceptable or non spec compliant?

Reply:
I was hoping for examples around what types of hosts, but if it's more heuristic-based today I don't think it's needed.

We only need origin here if we think people might want to use the port to determine whether or not to upgrade. I could see not wanting to upgrade ports other than the default port, for instance, but not sure if anyone is planning that. Maybe leave this as-is for now and revise if it comes up?

It would not constrain you just taking the host out of the origin and making decisions based on that. As you say, it makes it more flexible.

Reply:
Changed to origin. I left the new example in, let me know if you meant otherwise.